Fixes `mres` and `nres` in `free_resolution`

RafaelDavidMohr opened the pull request:
This PR will fix the `mres` and `nres` options in `free_resolution` using `prune_with_map`. Previously, using these options
might not return a minimal free resolution.

For using `prune_with_map` conveniently in `free_resolution`, I would like add a `minimal` keyword to `presentation`, if this is set to `true` we would return a minimal presentation, computed from the output of `prune_with_map`.

@wdecker Do we want a user-facing option to compute a minimal presentation? If not, then I will do this just internally.

RafaelDavidMohr committed:
adds `_presentation_minimal`

wdecker commented:
@RafaelDavidMohr As far as I can see, the new functionality is only used with `mres`. If everybody agrees, we do not need the proposed keyword. Or?

wdecker commented:
@jankoboehm userfacing function named `minimal_presentation`?

jankoboehm commented:
Yes we want the keyword. There are use-cases for both options.

RafaelDavidMohr committed:
adds `minimal` kw to presentation

RafaelDavidMohr committed:
use minimal kw in `free_resolution`

fingolfin commented on src/Modules/UngradedModules/Presentation.jl:
```diff
@@ -1,11 +1,15 @@
+                      minimal=false)
+  if minimal
+    return _presentation_minimal(SQ)
   if is_graded(SQ)
```
```suggestion
  elseif is_graded(SQ)
```

fingolfin commented on src/Modules/UngradedModules/Presentation.jl:
```diff
@@ -559,6 +563,50 @@ function prune_with_map(M::ModuleFP{T}) where {T<:MPolyRingElem{<:FieldElem}} #
+  proj_map = hom(F0, SQ_new, gens(SQ_new), check=false)
+  F0_to_SQ = compose(proj_map, phi)
+  F0_to_SQ.generators_map_to_generators = true
+  AbstractAlgebra.set_name!(F0, "$br_name^$(ngens(SQ.sub))")
```
Why do you ned to set this name here and below? Since F0 and F1 are free module anyway, couldn't this be done generally as part of the `show` methods for free modules?

As it is, this hardcodes the name of the basering, so if it changes later on, it will look ugly. Likewise if there is no name set, you hardcode `br` forever.

lgoettgens replied:
This is the current standard for presentations/resolutions/etc. and the same at other places. Thus I would keep it like this here right now and would instead think of some way to replace all of these instances by some AA printing magic in the future.

fingolfin replied:
Yeah if this is pervasive we should not try to deal with it in this PR. But perhaps we can open an issue as a reminder? 

Or perhaps we just add it to https://github.com/oscar-system/Oscar.jl/issues/2891

lgoettgens replied:
See https://github.com/oscar-system/Oscar.jl/issues/3730

RafaelDavidMohr commented:
@wdecker @jankoboehm do you have an example where `presentation` produces something non-minimal that I can use as a test for the minimal free resolution?

wdecker commented:
This example should do:

```julia
X = rational_d9_pi6();
Pn = ambient_coordinate_ring(X)
A = homogeneous_coordinate_ring(X)
n = ngens(Pn)-1
c = codim(X)
FA = free_resolution(A, algorithm = :fres)
C = Oscar.SimpleComplexWrapper(FA.C[0:first(range(FA.C))])
C_simp = simplify(C)
C_shift = shift(C_simp, c)
OmegaPn = graded_free_module(Pn, [n+1])
D = hom(C_shift, OmegaPn)
D_simp = simplify(D)
Z, inc = kernel(D_simp, 0)
B, inc_B = Oscar.boundary(D_simp, 0)
M = SubquoModule(D_simp[0], ambient_representatives_generators(Z), ambient_representatives_generators(B))
```

> Am 10.05.2024 um 17:13 schrieb Rafael Mohr ***@***.***>:
> 
> 
> @wdecker <https://github.com/wdecker> @jankoboehm <https://github.com/jankoboehm> do you have an example where presentation produces something non-minimal that I can use as a test for the minimal free resolution?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/oscar-system/Oscar.jl/pull/3700#issuecomment-2104784565>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AASSXETO6QBTAP236R4JSXLZBTPY7AVCNFSM6AAAAABHLCCC6KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMBUG44DINJWGU>.
> You are receiving this because you were mentioned.
>

wdecker commented:
one more:

S, (x0, x1, x2, x3, x4) = graded_polynomial_ring(GF(3), ["x0", "x1", "x2", "x3", "x4"]);
m = ideal(S, [x1^2+(-x1+x2+x3-x4)*x0, x1*x2+(x1-x3+x4)*x0, x1*x3+(-x1+x4+x0)*x0, x1*x4+(-x1+x3+x4-x0)*x0, x2^2+(x1-x2-x4-x0)*x0, x2*x3+(x1-x2+x3+x4-x0)*x0, x2*x4+(x1+x2-x3-x4-x0)*x0, x3^2+(x3+x4-x0)*x0,x3*x4+(-x3-x4+x0)*x0, x4^2+(x1+x3-x4-x0)*x0]);
Qm, _ = quo(S, m);
FQm = free_resolution(Qm)
M, _ = image(map(FQm, 1))



> Am 10.05.2024 um 17:13 schrieb Rafael Mohr ***@***.***>:
> 
> 
> @wdecker <https://github.com/wdecker> @jankoboehm <https://github.com/jankoboehm> do you have an example where presentation produces something non-minimal that I can use as a test for the minimal free resolution?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/oscar-system/Oscar.jl/pull/3700#issuecomment-2104784565>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AASSXETO6QBTAP236R4JSXLZBTPY7AVCNFSM6AAAAABHLCCC6KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMBUG44DINJWGU>.
> You are receiving this because you were mentioned.
>

RafaelDavidMohr committed:
if -> elseif

RafaelDavidMohr commented:
> one more: S, (x0, x1, x2, x3, x4) = graded_polynomial_ring(GF(3), ["x0", "x1", "x2", "x3", "x4"]); m = ideal(S, [x1^2+(-x1+x2+x3-x4)*x0, x1*x2+(x1-x3+x4)*x0, x1*x3+(-x1+x4+x0)*x0, x1*x4+(-x1+x3+x4-x0)*x0, x2^2+(x1-x2-x4-x0)*x0, x2*x3+(x1-x2+x3+x4-x0)*x0, x2*x4+(x1+x2-x3-x4-x0)*x0, x3^2+(x3+x4-x0)*x0,x3*x4+(-x3-x4+x0)*x0, x4^2+(x1+x3-x4-x0)*x0]); Qm, _ = quo(S, m); FQm = free_resolution(Qm) M, _ = image(map(FQm, 1))

Do we know the correct ranks for the minimal free resolution in this example so I can test?

wdecker commented:
In `_presentation_minimal` please catch border case of a free module before entering
`AbstractAlgebra.set_name!(F0, "$br_name^$(ngens(SQ.sub))")`

wdecker commented:
There is a problem with the print function:

julia> R, (x,y,z) = polynomial_ring(QQ, ["x", "y", "z"]);

julia> A = R[x 1 y; 2*x^2 z^2 3*y^2];

julia> M = cokernel(A);

julia> PM = presentation(M)
0 <---- M <---- R^3 <---- R^2

julia> MPM = _presentation_minimal(M)
0 <---- M <---- R^3 <---- R^1

julia> matrix(map(MPM,1))
[-2*x^2 + x*z^2   -3*y^2 + y*z^2]

julia> MPM[0]
Free module of rank 2 over R

RafaelDavidMohr commented:
> In `_presentation_minimal` please catch border case of a free module before entering `AbstractAlgebra.set_name!(F0, "$br_name^$(ngens(SQ.sub))")`

This case is caught in `presentation` already, and the user will only access `_presentation_minimal` via the `minimal` kw of presentation. Should I nonetheless treat this case seperately?

RafaelDavidMohr committed:
fix printing

RafaelDavidMohr commented:
> M = cokernel(A);

Fixed with https://github.com/oscar-system/Oscar.jl/pull/3700/commits/f2a684b09aa4e0b3d6f1bec15d624479e596465d

wdecker commented:
I am not sure what our policy here is. It may happen that a future developer may use this internal function and run into the border case. @fingolfin what do you think?

> Am 14.05.2024 um 11:54 schrieb Rafael Mohr ***@***.***>:
> 
> 
> In _presentation_minimal please catch border case of a free module before entering AbstractAlgebra.set_name!(F0, "$br_name^$(ngens(SQ.sub))")
> 
> This case is caught in presentation already, and the user will only access _presentation_minimal via the minimal kw of presentation. Should I nonetheless treat this case seperately?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/oscar-system/Oscar.jl/pull/3700#issuecomment-2109774630>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AASSXESK2FX6YF5GPGO34LDZCHNONAVCNFSM6AAAAABHLCCC6KVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMBZG43TINRTGA>.
> You are receiving this because you were mentioned.
>

lgoettgens commented on src/Modules/UngradedModules/Presentation.jl:
```diff
@@ -1,12 +1,16 @@
+  if minimal
+    return _presentation_minimal(SQ)
+  elseif is_graded(SQ)
     return _presentation_graded(SQ)
```
How do `minimal=true` and a graded module interact? Due to now calling `_presentation_minimal` here, the graded check is skipped. So `_presentation_minimal` would need to special-case graded modules somehow

RafaelDavidMohr replied:
Internally, `_presentation_minimal` uses `prune_with_map` which has to check for gradedness anyway. So `_presentation_minimal` does not have to do any extra checks.

lgoettgens replied:
Nice, didn't know that

fingolfin commented:
@wdecker I am afraid I don't understand what you are asking about. What is "this internal function", and which border case?

Happy to talk about it during triage today if anyone would like to.

lgoettgens commented:
> @wdecker I am afraid I don't understand what you are asking about. What is "this internal function", and which border case?
> 
> Happy to talk about it during triage today if anyone would like to.

I think I understood the issue and can try to explain it later in triage if wdecker is not there

RafaelDavidMohr committed:
minimize kernel in `_presentation_minimal`

RafaelDavidMohr committed:
min presentation test

RafaelDavidMohr commented:
I added the discussed minimization of the kernel in `_presentation_minimal`. I added it with an optional keyword, so that it can easily be integrated into `free_resolution`.

Currently this uses `prune_with_map`. I tried to use `Singular.mres` but this function seems to ignore the `length` argument. 

@wdecker Should I make a Singular issue regarding this?

ederc commented:
Can you give a concrete example where `Singular.mres` ignores the `length` argument? If I compute a free resolution with current Oscar master using `algorithm = :mres` and `length` set, it seems to compute up to the correct `length`.

wdecker reviewed (approved):
This is ready to go as soon as the tests are green. Please check tests.

RafaelDavidMohr committed:
fix tests

RafaelDavidMohr commented:
> Can you give a concrete example where `Singular.mres` ignores the `length` argument? If I compute a free resolution with current Oscar master using `algorithm = :mres` and `length` set, it seems to compute up to the correct `length`.

Maybe I'm misunderstanding the concept of length of a free resolution, but for example
```
julia> R, (x, y, z) = polynomial_ring(GF(65521), ["x", "y", "z"]);

julia> I = ideal(R, [x^2, x*y, y]);

julia> M = ideal_as_module(I);

julia> sf = Oscar.singular_module(ambient_free_module(M));

julia> sm = Singular.Module(base_ring(sf), [sf(repres(g)) for g in gens(M)]...);

julia> Singular.mres(sm, 1)
Singular resolution: R^1 <- R^2 <- R^1

julia> Singular.mres(sm, 2)
Singular resolution: R^1 <- R^2 <- R^1

julia> Singular.mres(sm, 3)
Singular resolution: R^1 <- R^2 <- R^1
```

wdecker commented:
@RafaelDavidMohr In your example, the resolution is short, so entering an increased length will not change anything:

julia> R, (x, y) = polynomial_ring(GF(65521), ["x", "y"]);

julia> I = ideal(R, [x^2, x*y, y]);

julia> M = ideal_as_module(I);

julia> FM = free_resolution(M, algorithm = :mres, length = 1)
Free resolution of M
R^3 <---- R^2
0         1

julia> FM = free_resolution(M, algorithm = :mres, length = 0)
Free resolution of M
R^3 <---- R^2 <---- 0
0         1         2


So why now does Singular print this:

julia> Singular.mres(sm, 1)
Singular resolution: R^1 <- R^2 <- R^1

I guess, this is due to the fact that you use `mres`. So you have to go one step further in order to minimalize. @hannes14 ?

RafaelDavidMohr committed:
fix tests
